src/utils/rdf.js

```diff
@@ -1,6 +1,75 @@
-import { Parser } from 'n3';
+// @ts-nocheck
+import { Parser, Store, DataFactory } from 'n3';
+import { newEngine } from '@comunica/actor-init-sparql-rdfjs';
 import { fromJsonldjsNode } from './claimgraph';
 
+/**
+ * Converts a a JSON-LD RDF object to N3 data type
+ * @param {object} node - JSON-LD RDF representation object
+ * @returns {any}
+ */
+export function toJsonldjsNode(node) {
+  if (node.DefaultGraph) {
+    return DataFactory.defaultGraph();
+  } else if (node.Iri) {
+    return DataFactory.namedNode(node.Iri);
+  } else if (node.Blank) {
+    return DataFactory.blankNode(node.Blank);
+  } else if (node.Literal) {
+    return DataFactory.literal(node.Literal.value, {
+      value: node.Literal.datatype,
+      language: node.Literal.language,
+    });
+  }
+
+  throw new Error(`Unable to determine node type: ${node}`);
+}
+
+/**
+ * Converts a claimgraph JSON representation into an N3 RDF store
+ * @param {object} claimgraph - JSON-LD RDF object
+ * @returns {Store}
+ */
+export function claimgraphToStore(claimgraph) {
+  const store = new Store();
+  claimgraph.forEach((quad) => {
+    const subject = quad[0];
+    const predicate = quad[1];
+    const object = quad[2];
+    const graph = quad.length > 3 ? toJsonldjsNode(quad[3]) : undefined;
+    store.addQuad(DataFactory.quad(
+      toJsonldjsNode(subject),
+      toJsonldjsNode(predicate),
+      toJsonldjsNode(object),
+      graph,
+    ));
+  });
+  return store;
+}
+
+/**
+ * Queries a claimgraph object and returns its bindings as an RDF term list
+ * @param {object} claimgraph - JSON-LD RDF object
+ * @param {string} query - SPARQL query string
+ * @param {object|null} engine - RDF query engine or null to auto-create
+ * @returns {array<any>}
+ */
+export async function queryClaimgraph(claimgraph, query, engine = null) {
+  // Create an N3 store from claimgraph JSON object
+  const store = typeof claimgraph === 'string' ? claimgraph : claimgraphToStore(claimgraph);
+
+  // Query the engine, if querying multiple times user should
+  // pass engine parameter for optimal performance
+  const myEngine = engine || newEngine();
+  const result = await myEngine.query(query, { sources: [store] });
+
+  // Get bindings from query
+  const bindings = await result.bindings();
+
+  // Convert bindings to claimgraph format using lookupNext variable
+  return bindings.map((binding) => fromJsonldjsNode(binding.get('?lookupNext')));
```

Comment:
How will this behave for a query that is invalid because it has no the `?lookupNext` variable.

Reply:
I was thinking about this, and I think it would return potentially null/undefined in the array. I wasn't sure if its good to hardcode this, maybe exposing a parameter with `?lookupNext` as a default value would make more sense. That way a user can choose to get their own bindings if they wish, rather than this method just resulting in an error if not found

Reply:
As a reminder, this method doesn't need to be exposed to users. It's only needed for use by the crawler.

Users will be able to provide the sparql query, but they don't need to call this function directly.

Reply:
To avoid hard-coding `lookupNext` this function could just return `bindings` and let the crawler do the rest.

Reply:
if we're just returning bindings then this method to me would seem useless, as its basically just a wrapper around query and bindings return. do you have anything against using a parameter to define which binding to get? im happy with `lookupNext` as a default value

Reply:
alternatively, we can just assume user has to use lookupNext in their query and then if we cant find any binding with that value defined we can throw an error in this method. it's hard to determine the best approach with no crawler implementation, feels a bit silo'd

Reply:
I think any of  the proposed solutions works, including how it is now.

Thanks for mentioning that this feels silo'd. I think you are right and it's probably a pain to have to bounce stuff off me instead of me explaining my plans in more detail up-front.